File: android-kmp/shared/network/src/commonMain/kotlin/com/embytok/network/client/EmbyClient.kt
package com.embytok.network.client

import com.embytok.domain.client.MediaClient
import com.embytok.domain.model.EmbyItem
import com.embytok.domain.model.EmbyLibrary
import com.embytok.domain.model.MediaSource
import com.embytok.domain.model.SubtitleTrack
import com.embytok.domain.model.UserData
import com.embytok.network.defaultHttpClient
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Emby 服务器原生客户端（Ktor + JSON）。
 *
 * API 约定：
 *   - HTTP Header: `X-Emby-Token` / `X-Emby-Authorization`
 *   - 认证: POST /Users/AuthenticateByName
 *   - 库列表: GET /Library/VirtualFolders  或  GET /Library/MediaFolders
 *   - 条目列表: GET /Users/{userId}/Items?ParentId={libraryId}&Recursive=true
 *   - 详情: GET /Users/{userId}/Items/{id}
 */
class EmbyClient(
    private val baseUrl: String,
    apiKey: String? = null,
    userId: String? = null,
    private val httpClient: HttpClient = defaultHttpClient()
) : MediaClient {

    // 可变字段：authenticate 成功后会用 AccessToken / User.Id 填充
    @Volatile
    private var currentApiKey: String? = apiKey?.ifBlank { null }

    @Volatile
    private var currentUserId: String? = userId?.ifBlank { null }

    private val authHeaderValue =
        "MediaBrowser Client=\"EmbyTok\", Device=\"Android\", DeviceId=\"embytok\", Version=\"1.0.0\""

    private fun apiPath(path: String): String = "${baseUrl.trimEnd('/')}/${path.trimStart('/')}"

    private fun requireUserId(): String =
        currentUserId ?: throw IllegalStateException("需要先登录获取 userId")

    private fun requireApiKey(): String =
        currentApiKey ?: throw IllegalStateException("需要 API Key 或先登录")

    // ===== 认证 =====

    override suspend fun ping(): Result<String> = runCatching {
        val token = currentApiKey
        val resp = httpClient.get(apiPath("System/Info")) {
            if (token != null) header("X-Emby-Token", token)
            header("X-Emby-Authorization", authHeaderValue)
        }.body<EmbySystemInfo>()
        resp.Id?.ifBlank { "emby-server" } ?: "emby-server"
    }

    override suspend fun authenticate(username: String, password: String): Result<String> = runCatching {
        val body = EmbyAuthBody(username, password, "EmbyTok", "Android", "embytok", "1.0.0")
        val resp = httpClient.post(apiPath("Users/AuthenticateByName")) {
            header("X-Emby-Authorization", authHeaderValue)
            contentType(ContentType.Application.Json)
            setBody(body)
        }.body<EmbyAuthJsonResponse>()
        // 认证成功：保存 AccessToken 与 userId，供后续请求使用
        currentApiKey = resp.AccessToken
        currentUserId = resp.User.Id
        resp.User.Id
    }

    /** 当前已登录的 userId（可用于构造 ServerConfig） */
    fun currentUserId(): String? = currentUserId

    /** 当前已登录的 token（可用于构造 ServerConfig） */
    fun currentToken(): String? = currentApiKey

    // ===== 媒体库 =====

    override suspend fun getLibraries(): List<EmbyLibrary> {
        return runCatching {
            val token = requireApiKey()
            val resp = httpClient.get(apiPath("Library/VirtualFolders")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
            }.body<EmbyItemsResponse<EmbyLibraryJson>>()
            resp.Items.map { json ->
                EmbyLibrary(
                    Id = json.Id,
                    Name = json.Name,
                    CollectionType = json.CollectionType
                )
            }
        }.getOrDefault(emptyList())
    }

    // ===== 视频列表 =====

    override suspend fun getLibraryItems(libraryId: String): List<EmbyItem> {
        val uid = requireUserId()
        val token = requireApiKey()
        return runCatching {
            val resp = httpClient.get(apiPath("Users/$uid/Items")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                parameter("ParentId", libraryId)
                parameter("Recursive", "true")
                parameter("IncludeItemTypes", "Movie,Episode,Video")
                parameter("Fields", "MediaSources,MediaStreams,Overview,ProductionYear,Width,Height,RuntimeTicks,UserData")
                parameter("SortBy", "DateCreated,SortName")
                parameter("SortOrder", "Descending")
                parameter("Limit", "300")
            }.body<EmbyItemsResponse<EmbyItemJson>>()
            resp.Items.map { it.toEmbyItem() }
        }.getOrDefault(emptyList())
    }

    override suspend fun getLatestItems(libraryId: String, limit: Int): List<EmbyItem> {
        val uid = requireUserId()
        val token = requireApiKey()
        return runCatching {
            val resp = httpClient.get(apiPath("Users/$uid/Items/Latest")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                parameter("ParentId", libraryId)
                parameter("Limit", limit.toString())
                parameter("Fields", "MediaSources,Overview,ProductionYear,Width,Height,RuntimeTicks,UserData")
            }.body<List<EmbyItemJson>>()
            resp.map { it.toEmbyItem() }
        }.getOrDefault(emptyList())
    }

    override suspend fun getFavoriteItems(libraryId: String): List<EmbyItem> {
        val uid = requireUserId()
        val token = requireApiKey()
        return runCatching {
            val resp = httpClient.get(apiPath("Users/$uid/Items")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                parameter("ParentId", libraryId)
                parameter("Recursive", "true")
                parameter("Filters", "IsFavorite")
                parameter("IncludeItemTypes", "Movie,Episode,Video")
                parameter("Fields", "MediaSources,Overview,ProductionYear,Width,Height,RuntimeTicks,UserData")
                parameter("Limit", "300")
            }.body<EmbyItemsResponse<EmbyItemJson>>()
            resp.Items.map { it.toEmbyItem() }
        }.getOrDefault(emptyList())
    }

    // ===== 收藏 / 观看状态 =====

    override suspend fun toggleFavorite(itemId: String): Boolean {
        val uid = requireUserId()
        val token = requireApiKey()
        return runCatching {
            httpClient.post(apiPath("Users/$uid/Favorites/$itemId")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                contentType(ContentType.Application.Json)
            }
            true
        }.getOrDefault(false)
    }

    override suspend fun markAsWatched(itemId: String) {
        val uid = requireUserId()
        val token = requireApiKey()
        runCatching {
            httpClient.post(apiPath("Users/$uid/PlayedItems/$itemId")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                contentType(ContentType.Application.Json)
            }
        }
    }

    // ===== 字幕 =====

    override suspend fun getSubtitles(itemId: String): List<SubtitleTrack> {
        val detail = runCatching {
            val uid = requireUserId()
            val token = requireApiKey()
            httpClient.get(apiPath("Users/$uid/Items/$itemId")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                parameter("Fields", "MediaSources,MediaStreams")
            }.body<EmbyItemJson>()
        }.getOrNull() ?: return emptyList()

        val tracks = mutableListOf<SubtitleTrack>()
        detail.MediaSources?.forEach { source ->
            source.MediaStreams?.filter { it.Type == "Subtitle" }?.forEach { stream ->
                val idx = stream.Index
                val src = apiPath("Videos/$itemId/${source.Id}/Subtitles/$idx/0/WebVTT")
                tracks += SubtitleTrack(
                    id = "$itemId-${source.Id}-$idx",
                    label = stream.DisplayTitle,
                    srclang = stream.Language,
                    src = src,
                    type = "vtt"
                )
            }
        }
        return tracks
    }

    // ===== 搜索 =====

    override suspend fun searchItems(query: String, limit: Int): List<EmbyItem> {
        val uid = requireUserId()
        val token = requireApiKey()
        return runCatching {
            val resp = httpClient.get(apiPath("Users/$uid/Items")) {
                header("X-Emby-Token", token)
                header("X-Emby-Authorization", authHeaderValue)
                parameter("SearchTerm", query)
                parameter("Recursive", "true")
                parameter("IncludeItemTypes", "Movie,Episode,Video")
                parameter("Fields", "MediaSources,MediaStreams,Overview,ProductionYear,Width,Height,RuntimeTicks,UserData")
                parameter("Limit", limit.toString())
            }.body<EmbyItemsResponse<EmbyItemJson>>()
            resp.Items.map { it.toEmbyItem() }
        }.getOrDefault(emptyList())
    }

    // ===== 播放地址 =====

    override fun buildVideoStreamUrl(itemId: String, mediaSourceId: String?): String {
        val sourceId = mediaSourceId ?: itemId
        val token = currentApiKey.orEmpty()
        return apiPath("Videos/$itemId/stream.mp4?Static=true&MediaSourceId=$sourceId&X-Emby-Token=$token")
    }

    // ===== 图片 URL 构建 =====

    /**
     * 构造 Emby 图片 URL。
     *
     * Emby 图片 API 格式：
     *   /Items/{itemId}/Images/{imageTag}?MaxWidth={w}&MaxHeight={h}&X-Emby-Token={token}
     *
     * 常见 imageTag 值：
     *   - "Primary"   主封面图（海报）
     *   - "Thumb"     缩略图（横向）
     *   - "Backdrop"  背景/场景图
     *   - "Logo"      标题 Logo
     */
    override fun buildImageUrl(
        itemId: String,
        imageTag: String,
        maxWidth: Int,
        maxHeight: Int
    ): String {
        val token = currentApiKey.orEmpty()
        return apiPath("Items/$itemId/Images/$imageTag") +
                "?MaxWidth=$maxWidth&MaxHeight=$maxHeight&X-Emby-Token=$token"
    }

    // ============ 内部 JSON 模型 ============

    @Serializable
    private data class EmbySystemInfo(
        val Id: String? = null,
        val ServerName: String? = null
    )

    @Serializable
    private data class EmbyAuthBody(
        val Username: String,
        val Pw: String,
        val App: String,
        val Device: String,
        val DeviceId: String,
        val Version: String
    )

    @Serializable
    private data class EmbyAuthJsonResponse(
        val User: EmbyUserJson,
        val AccessToken: String,
        val ServerId: String
    )

    @Serializable
    private data class EmbyUserJson(val Id: String, val Name: String)

    @Serializable
    private data class EmbyItemsResponse<T>(val Items: List<T>, val TotalRecordCount: Int)

    @Serializable
    private data class EmbyLibraryJson(
        val Id: String,
        val Name: String,
        val CollectionType: String? = null
    )

    @Serializable
    private data class EmbyItemJson(
        val Id: String,
        val Name: String,
        val Type: String = "Video",
        val Overview: String? = null,
        val ProductionYear: Int? = null,
        val RunTimeTicks: Long? = null,
        val Width: Int? = null,
        val Height: Int? = null,
        val ParentId: String? = null,
        val UserData: EmbyUserDataJson? = null,
        val MediaSources: List<EmbyMediaSourceJson>? = null,
        val IndexNumber: Int? = null,
        val ParentIndexNumber: Int? = null,
        val SeriesName: String? = null
    ) {
        fun toEmbyItem(): EmbyItem = EmbyItem(
            Id = Id,
            Name = Name,
            Type = Type,
            Overview = Overview,
            ProductionYear = ProductionYear,
            RunTimeTicks = RunTimeTicks,
            Width = Width,
            Height = Height,
            ParentId = ParentId,
            UserData = UserData?.let {
                UserData(
                    IsFavorite = it.IsFavorite ?: false,
                    PlaybackPositionTicks = it.PlaybackPositionTicks ?: 0,
                    PlayCount = it.PlayCount ?: 0,
                    Played = it.Played ?: false
                )
            },
            MediaSources = MediaSources?.map { ms ->
                MediaSource(
                    Id = ms.Id,
                    Path = ms.Path,
                    Container = ms.Container,
                    Bitrate = ms.Bitrate,
                    SupportsDirectPlay = ms.SupportsDirectPlay ?: true,
                    SupportsDirectStream = ms.SupportsDirectStream,
                    SupportsTranscoding = ms.SupportsTranscoding,
                    Protocol = ms.Protocol,
                    MediaStreams = ms.MediaStreams?.map { s ->
                        com.embytok.domain.model.MediaStream(
                            Index = s.Index,
                            Type = s.Type,
                            Codec = s.Codec,
                            DisplayTitle = s.DisplayTitle,
                            Language = s.Language,
                            IsDefault = s.IsDefault,
                            IsExternal = s.IsExternal ?: false,
                            Path = s.Path
                        )
                    }
                )
            },
            IndexNumber = IndexNumber,
            ParentIndexNumber = ParentIndexNumber,
            SeriesName = SeriesName
        )
    }

    @Serializable
    private data class EmbyUserDataJson(
        val IsFavorite: Boolean? = null,
        val PlaybackPositionTicks: Long? = null,
        val PlayCount: Int? = null,
        val Played: Boolean? = null
    )

    @Serializable
    private data class EmbyMediaSourceJson(
        val Id: String,
        val Path: String? = null,
        val Container: String? = null,
        val Bitrate: Int? = null,
        val Protocol: String? = null,
        val SupportsDirectPlay: Boolean? = null,
        val SupportsDirectStream: Boolean? = null,
        val SupportsTranscoding: Boolean? = null,
        val MediaStreams: List<EmbyMediaStreamJson>? = null
    )

    @Serializable
    private data class EmbyMediaStreamJson(
        val Index: Int = 0,
        val Type: String? = null,
        val Codec: String? = null,
        val DisplayTitle: String? = null,
        val Language: String? = null,
        val IsDefault: Boolean? = null,
        val IsExternal: Boolean? = null,
        val Path: String? = null
    )
}

File: services/EmbyClient.ts
import { MediaClient } from './MediaClient';
import {
  EmbyItem,
  EmbyLibrary,
  FeedType,
  ServerConfig,
  VideoResponse,
  OrientationMode,
  SubtitleTrack,
} from '../types';
import { getDeviceId } from '../utils/device';

const DEVICE_ID = getDeviceId();

export class EmbyClient extends MediaClient {
  private getHeaders() {
    return {
      'Content-Type': 'application/json',
      'X-Emby-Authorization': `MediaBrowser Client="EmbyTok Web", Device="Web Browser", DeviceId="${DEVICE_ID}", Version="1.0.0", Token="${this.config.token}"`,
      'X-Emby-Token': this.config.token,
      'X-MediaBrowser-Token': this.config.token,
    };
  }

  private getCleanUrl() {
    return this.config.url.replace(/\/$/, '');
  }

  async authenticate(username: string, password: string): Promise<ServerConfig> {
    const response = await fetch(`${this.getCleanUrl()}/Users/AuthenticateByName`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify({ Username: username, Pw: password }),
    });
    if (!response.ok) throw new Error('Emby Authentication failed');
    const data = await response.json();
    return {
      url: this.config.url,
      username: data.User.Name,
      userId: data.User.Id,
      token: data.AccessToken,
      serverType: 'emby',
    };
  }

  async getLibraries(): Promise<EmbyLibrary[]> {
    const response = await fetch(`${this.getCleanUrl()}/Users/${this.config.userId}/Views`, {
      headers: this.getHeaders(),
    });
    const data = await response.json();
    return data.Items || [];
  }

  // 仅为 TV 首页增加此接口，不干扰原有逻辑
  async getResumeItems(): Promise<EmbyItem[]> {
    const params = new URLSearchParams({
      Recursive: 'true',
      Fields: 'PrimaryImageAspectRatio,BasicSyncInfo,ProductionYear,UserData',
      ImageTypeLimit: '1',
      EnableImageTypes: 'Primary,Backdrop,Thumb',
      MediaTypes: 'Video',
      Limit: '12',
    });
    const response = await fetch(
      `${this.getCleanUrl()}/Users/${this.config.userId}/Items/Resume?${params.toString()}`,
      { headers: this.getHeaders() }
    );
    const data = await response.json();
    return (data.Items || []).map((i: any) => ({ ...i, Name: this.formatItemName(i) }));
  }

  private formatItemName(item: any): string {
    if (item.Type === 'Episode') {
      const index =
        item.IndexNumber !== undefined ? String(item.IndexNumber).padStart(2, '0') : '--';
      const season =
        item.ParentIndexNumber !== undefined
          ? `S${String(item.ParentIndexNumber).padStart(2, '0')}`
          : '';
      return `${season}E${index}. ${item.Name}`;
    }
    return item.Name || '未命名';
  }

  private applyOrientationFilter(items: any[], mode: OrientationMode): any[] {
    if (mode === 'both') return items;
    return items.filter((item) => {
      const isNavFolder = ['Series', 'Season', 'Folder', 'CollectionFolder', 'BoxSet'].includes(
        item.Type
      );
      if (isNavFolder) return true;
      const w = item.Width || 0;
      const h = item.Height || 0;
      if (w === 0 || h === 0) return true;
      if (mode === 'vertical') return h >= w * 0.8;
      if (mode === 'horizontal') return w > h;
      return true;
    });
  }

  async getVideos(
    navParentId: string | undefined,
    library: EmbyLibrary | null,
    feedType: FeedType,
    skip: number,
    limit: number,
    orientationMode: OrientationMode,
    includeIds?: string
  ): Promise<VideoResponse> {
    const libraryName = library ? library.Name : '收藏';

    // 严格遵循 backup 中的 Playlist 逻辑
    if (feedType === 'favorites') {
      const playlistItems = await this.getTokPlaylistItemsInternal(libraryName);
      const filtered = this.applyOrientationFilter(playlistItems, orientationMode);
      const paged = filtered.reverse().slice(skip, skip + limit);
      return { items: paged, nextStartIndex: skip + limit, totalCount: filtered.length };
    }

    const params = new URLSearchParams({
      Fields:
        'MediaSources,Width,Height,Overview,UserData,SeriesName,ParentIndexNumber,IndexNumber,Type',
      Limit: (limit * 2).toString(),
      StartIndex: skip.toString(),
      EnableImageTypes: 'Primary,Backdrop,Banner,Thumb',
      _t: Date.now().toString(),
    });

    if (includeIds && !navParentId && !library) {
      params.append('ParentIds', includeIds);
    }

    if (navParentId) {
      params.append('ParentId', navParentId);
      params.append('Recursive', 'false');
      params.append('SortBy', 'SortName');
      params.append('IncludeItemTypes', 'Movie,Video,Episode,Folder,BoxSet,Series,Season');
    } else {
      if (library) {
        params.append('ParentId', library.Id);
        const collectionType = (library.CollectionType || '').toLowerCase();
        if (collectionType === 'tvshows' || collectionType === 'show') {
          params.append('IncludeItemTypes', 'Series');
        } else if (collectionType === 'folders') {
          params.append('IncludeItemTypes', 'Movie,Video,Episode,Folder,BoxSet');
        } else {
          params.append('IncludeItemTypes', 'Movie,Video,Episode');
        }
      } else {
        params.append('IncludeItemTypes', 'Movie,Video,Episode');
      }
      params.append('Recursive', 'true');
      params.append('SortBy', feedType === 'random' ? 'Random' : 'DateCreated');
      params.append('SortOrder', 'Descending');
    }

    const response = await fetch(
      `${this.getCleanUrl()}/Users/${this.config.userId}/Items?${params.toString()}`,
      { headers: this.getHeaders() }
    );
    const data = await response.json();
    const rawItems = data.Items || [];
    const filteredItems = this.applyOrientationFilter(rawItems, orientationMode);

    const items = filteredItems.map((item: any) => ({
      ...item,
      Name: this.formatItemName(item),
      UserData: item.UserData
        ? {
            ...item.UserData,
            Played: item.UserData.Played || false,
            PlaybackPositionTicks: item.UserData.PlaybackPositionTicks || 0,
            LastPlayedDate: item.UserData.LastPlayedDate,
          }
        : undefined,
    }));

    return {
      items,
      nextStartIndex: skip + rawItems.length,
      totalCount: data.TotalRecordCount || 0,
    };
  }

  getVideoUrl(item: EmbyItem, mode: 'direct' | 'transcode' | 'fallback' = 'direct'): string {
    const playSessionId = Date.now();
    const baseUrl = this.getCleanUrl();

    if (mode === 'direct' && item.MediaSources && item.MediaSources.length > 0) {
      // 直接播放模式：优先从MediaSources中找到可以直接播放的媒体源
      const directSource = item.MediaSources.find((m) => m.SupportsDirectPlay && m.Path);
      if (directSource) {
        return `${baseUrl}/Videos/${item.Id}/stream?Static=true&MediaSourceId=${directSource.Id}&PlaySessionId=${playSessionId}&api_key=${this.config.token}`;
      }

      // 如果没有找到直接播放源，也尝试直接流
      return `${baseUrl}/Videos/${item.Id}/stream?Static=true&MediaSourceId=${item.Id}&PlaySessionId=${playSessionId}&RequireAvc=false&RequireNonAnamorphic=false&MaxWidth=3840&MaxHeight=2160&api_key=${this.config.token}`;
    }

    if (mode === 'transcode') {
      // 转码模式：让Emby服务器转码为兼容格式
      return `${baseUrl}/Videos/${item.Id}/master.m3u8?MediaSourceId=${item.Id}&PlaySessionId=${playSessionId}&VideoCodec=h264&AudioCodec=aac&MaxWidth=1920&MaxHeight=1080&VideoBitrate=4000000&AudioBitrate=192000&api_key=${this.config.token}`;
    }

    // 备用模式：更保守的转码设置
    return `${baseUrl}/Videos/${item.Id}/stream?MediaSourceId=${item.Id}&PlaySessionId=${playSessionId}&VideoCodec=h264&AudioCodec=aac&MaxWidth=1280&MaxHeight=720&VideoBitrate=2000000&AudioBitrate=128000&api_key=${this.config.token}`;
  }

  getImageUrl(itemId: string, tag?: string, type: 'Primary' | 'Backdrop' = 'Primary'): string {
    if (!tag) return '';
    // 补全 api_key 确保 TV 端加载正常
    return `${this.getCleanUrl()}/Items/${itemId}/Images/${type}?maxWidth=800&tag=${tag}&quality=90&api_key=${this.config.token}`;
  }

  // --- 恢复 Playlist 原始实现 ---
  private async getTokPlaylistId(libraryName: string): Promise<string> {
    const playlistName = `Tok-${libraryName}`;
    const searchRes = await fetch(
      `${this.getCleanUrl()}/Users/${this.config.userId}/Items?IncludeItemTypes=Playlist&Recursive=true`,
      { headers: this.getHeaders() }
    );
    const searchData = await searchRes.json();
    const existing = searchData.Items?.find((i: any) => i.Name === playlistName);
    if (existing) return existing.Id;
    const createRes = await fetch(
      `${this.getCleanUrl()}/Playlists?Name=${playlistName}&UserId=${this.config.userId}`,
      { method: 'POST', headers: this.getHeaders() }
    );
    const createData = await createRes.json();
    return createData.Id;
  }

  private async getTokPlaylistItemsInternal(libraryName: string): Promise<EmbyItem[]> {
    try {
      const pid = await this.getTokPlaylistId(libraryName);
      const response = await fetch(
        `${this.getCleanUrl()}/Playlists/${pid}/Items?UserId=${this.config.userId}&Fields=MediaSources,Width,Height,Overview,UserData`,
        { headers: this.getHeaders() }
      );
      const data = await response.json();
      return data.Items || [];
    } catch (e) {
      return [];
    }
  }

  async getFavorites(libraryName: string): Promise<Set<string>> {
    const items = await this.getTokPlaylistItemsInternal(libraryName);
    return new Set(items.map((i) => i.Id));
  }

  async toggleFavorite(itemId: string, isFavorite: boolean, libraryName: string): Promise<void> {
    const pid = await this.getTokPlaylistId(libraryName);
    if (!isFavorite) {
      await fetch(
        `${this.getCleanUrl()}/Playlists/${pid}/Items?Ids=${itemId}&UserId=${this.config.userId}`,
        { method: 'POST', headers: this.getHeaders() }
      );
    } else {
      const itemsRes = await fetch(
        `${this.getCleanUrl()}/Playlists/${pid}/Items?UserId=${this.config.userId}`,
        { headers: this.getHeaders() }
      );
      const entry = (await itemsRes.json()).Items.find((i: any) => i.Id === itemId);
      if (entry?.PlaylistItemId) {
        await fetch(
          `${this.getCleanUrl()}/Playlists/${pid}/Items?EntryIds=${entry.PlaylistItemId}`,
          { method: 'DELETE', headers: this.getHeaders() }
        );
      }
    }
  }

  async deleteItem(itemId: string): Promise<void> {
    const response = await fetch(
      `${this.getCleanUrl()}/Items/${itemId}?api_key=${this.config.token}`,
      {
        method: 'DELETE',
        headers: this.getHeaders(),
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to delete item: ${response.status}`);
    }
  }

  async searchItems(query: string): Promise<EmbyItem[]> {
    const params = new URLSearchParams({
      SearchTerm: query,
      IncludeItemTypes: 'Movie,Video,Episode,Series',
      Recursive: 'true',
      Fields:
        'MediaSources,Width,Height,Overview,UserData,SeriesName,ParentIndexNumber,IndexNumber,Type',
      Limit: '50',
    });

    const response = await fetch(
      `${this.getCleanUrl()}/Users/${this.config.userId}/Items?${params.toString()}`,
      {
        headers: this.getHeaders(),
      }
    );

    const data = await response.json();
    return (data.Items || []).map((item: any) => ({
      ...item,
      Name: this.formatItemName(item),
      UserData: item.UserData
        ? {
            ...item.UserData,
            Played: item.UserData.Played || false,
            PlaybackPositionTicks: item.UserData.PlaybackPositionTicks || 0,
            LastPlayedDate: item.UserData.LastPlayedDate,
          }
        : undefined,
    }));
  }

  async getSubtitleTracks(itemId: string): Promise<SubtitleTrack[]> {
    try {
      const params = new URLSearchParams({
        Fields: 'MediaSources',
      });

      const response = await fetch(
        `${this.getCleanUrl()}/Users/${this.config.userId}/Items/${itemId}?${params.toString()}`,
        {
          headers: this.getHeaders(),
        }
      );

      const data = await response.json();
      const mediaSources = data.MediaSources || [];

      const subtitleTracks: SubtitleTrack[] = [];
      mediaSources.forEach((source: any) => {
        if (source.MediaStreams) {
          source.MediaStreams.filter((stream: any) => stream.Type === 'Subtitle').forEach(
            (stream: any) => {
              let url: string | undefined;
              if (stream.IsExternal && stream.Path) {
                url = `${this.getCleanUrl()}/Videos/${itemId}/${stream.Index}/Stream?api_key=${this.config.token}`;
              } else if (stream.Codec === 'srt' || stream.Codec === 'vtt') {
                url = `${this.getCleanUrl()}/Videos/${itemId}/${stream.Index}/Stream?api_key=${this.config.token}`;
              }

              subtitleTracks.push({
                id: `${itemId}_${stream.Index}`,
                label:
                  stream.DisplayTitle || stream.Language || `字幕 ${subtitleTracks.length + 1}`,
                language: stream.Language || 'und',
                isDefault: stream.IsDefault || false,
                codec: stream.Codec,
                isExternal: stream.IsExternal || false,
                url,
              });
            }
          );
        }
      });

      return subtitleTracks;
    } catch (error) {
      console.error('Failed to get subtitle tracks:', error);
      return [];
    }
  }
}
